### PR TITLE
Actions dag 2

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Continuation of #16519

### CODE REVIEW

File: src/Interpreters/ExpressionActions.cpp [hunk 1]
```diff
@@ -1702,70 +1024,250 @@ std::string ActionsDAG::dumpNames() const
...
+            if (visited_nodes.count(child) == 0)
+            {
+                stack.push(child);
+                visited_nodes.insert(child);
+            }
         }
     }
 
-    for (auto & node : nodes)
+    nodes.remove_if([&](const Node & node) { return visited_nodes.count(&node) == 0; });
```
Comment (alesapin) [inline src/Interpreters/ExpressionActions.cpp:1094]: Seems like we need to erase removed nodes:
```
 A call to remove is typically followed by a call to a container's erase method, which erases the unspecified values and reduces the physical size of the container to match its new logical size. 
```

File: src/Interpreters/ExpressionActions.h [hunk 2]
```diff
@@ -184,102 +102,233 @@ class ActionsDAG
...
+    void removeUnusedActions();
+    void addAliases(const NamesWithAliases & aliases, std::vector<Node *> & result_nodes);
+
+    void compileFunctions();
 };
 
-using ActionsDAGPtr = std::shared_ptr<ActionsDAG>;
 
 /** Contains a sequence of actions on the block.
   */
```
Comment (alesapin) [inline src/Interpreters/ExpressionActions.h:260]: How ExpressionActions related to ActionsDAG? Seems like comment slightly outdated.

File: src/Interpreters/ExpressionActions.h [hunk 3]
```diff
@@ -184,102 +102,233 @@ class ActionsDAG
...
 class ExpressionActions
 {
 public:
```
Comment (alesapin) [inline src/Interpreters/ExpressionActions.h:263]: Maybe move DAG and ExpressionActions to separate files?

**Review by alesapin — COMMENTED** [review]
One of the most complex parts of ClickHouse's query execution stages was rewritten in this PR. It's a huge amount of work and honestly, I don't understand how it's possible to do it at once :) But tests pass and it seems extremely cool. 

This code was quite complicated mostly because it's contained really complex logic, not because of design flaws. But now it becomes completely different and there no explanation why we do this. There are quite a few comments in the code and we don't have anything to build a big picture -- what is going here.

Need to add an overview -- how DAG works, how actions are executed, why do we use this representation at all, and so on. Also need to add more comments to methods. I bet you, in the current form nobody will be able to understand these changes without direct questions. Again not because of new design flaws, but because of really complicated logic. Comments are the only way to make these code more clear.

Also, let's try to run tests with `compile_expression=1` by default.

File: src/Interpreters/ExpressionActions.cpp [hunk 4]
```diff
@@ -53,614 +54,268 @@ namespace ErrorCodes
...
+    if (!block)
+        block.insert({DataTypeUInt8().createColumnConst(num_rows, 0), std::make_shared<DataTypeUInt8>(), "_dummy"});
 }
 
-void ExpressionAction::execute(Block & block, bool dry_run) const
+void ExpressionActions::executeAction(const Action & action, ExecutionContext & execution_context, bool dry_run)
 {
-    size_t input_rows_count = block.rows();
-
-    if (type == REMOVE_COLUMN || type == COPY_COLUMN)
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Interpreters/ExpressionActions.cpp:346]: There is no `REMOVE_COLUMN` action any more.
For `COPY_COLUMN`, which now is names as `ALIAS` (and for other actions too), we
1. Check that column is in DAG when action is added
2. Check that inputs are in block when execute actions

File: src/Interpreters/ExpressionActions.cpp (hunk 1 above)
Comment (KochetovNicolai) [reply src/Interpreters/ExpressionActions.cpp:1094]: It is remove for `std::list`
```
The similarly-named container member functions list::remove, list::remove_if, forward_list::remove, and forward_list::remove_if erase the removed elements.
```

File: src/Interpreters/ExpressionActions.h [hunk 5]
```diff
@@ -373,43 +393,42 @@ struct ExpressionActionsChain
...
 
         explicit ExpressionActionsStep(ActionsDAGPtr actions_, Names required_output_ = Names())
             : Step(std::move(required_output_))
-            , actions_dag(std::move(actions_))
+            , actions(std::move(actions_))
         {
         }
 
-        const NamesAndTypesList & getRequiredColumns() const override
+        NamesAndTypesList getRequiredColumns() const override
```
[the remark this answers is not in the document]
Comment (KochetovNicolai) [reply src/Interpreters/ExpressionActions.h:408]: Cause `ActionsDAG` does not support direct list of required columns.

File: src/Interpreters/ExpressionJIT.cpp [hunk 6]
```diff
@@ -565,71 +586,166 @@ static bool isCompilable(const IFunctionBase & function)
...
+        size_t next_child_to_visit = 0;
+    };
+
+    std::stack<Frame> stack;
+    stack.push(Frame{.node = root});
+
+    while (!stack.empty())
+    {
+        auto & frame = stack.top();
+        bool is_const = isCompilableConstant(*frame.node);
```
Comment (alesapin) [inline src/Interpreters/ExpressionJIT.cpp:619]: Seems like compile expressions should be fixed now.

File: src/Interpreters/ExpressionJIT.cpp [hunk 7]
```diff
@@ -565,71 +586,166 @@ static bool isCompilable(const IFunctionBase & function)
...
+        hash.update(node.result_type->getName());
+
+        switch (node.type)
+        {
+            case CompileNode::NodeType::CONSTANT:
             {
-                Names columns = actions[i].getNeededColumns();
-                for (const auto & column : columns)
-                    current_dependents[column].emplace();
+                typeid_cast<const ColumnConst *>(node.column.get())->getDataColumn().updateHashWithValue(0, hash);
```
Comment (alesapin) [inline src/Interpreters/ExpressionJIT.cpp:722]: :+1:

**Review by alesapin — APPROVED** [review]

**Comment by alesapin** [comment]
Also, we need to remove commented lines?

**Comment by KochetovNicolai** [comment]
I will remove commented code in text pr.

**Pull request merged**